### PR TITLE
MM-23722 modify team privacy

### PR DESCRIPTION
**Pull request by wiggin77**
#### Summary

This PR implements a new API endpoint which modifies a team's privacy setting.  This API mimics server CLI API "team modify".

"public" changes team type to `model.TEAM_OPEN` and AllowOpenInvitation to `true`
"private" changes team type to `model.TEAM_INVITE` and AllowOpenInvitation to `false`

Includes units tests.

API docs here:  https://github.com/mattermost/mattermost-api-reference/pull/511

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-23722

### CODE REVIEW

File: api4/team.go [hunk 1]
```diff
@@ -231,6 +232,55 @@ func patchTeam(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	if team, err := c.App.GetTeam(c.Params.TeamId); err == nil {
+		auditRec.AddMeta("team", team)
+	}
+
+	var teamType string
+	var openInvite bool
+	switch c.Params.Privacy {
```
Comment (saturninoabril) [inline api4/team.go:262]: In [channel privacy API](https://api.mattermost.com/#tag/channels/paths/~1channels~1{channel_id}~1privacy/put), it's on the request body.  Would it make sense to do the same here?

File: api4/team_test.go [hunk 2]
```diff
@@ -459,6 +459,68 @@ func TestPatchTeamSanitization(t *testing.T) {
...
+			}
+
+			team, err := th.App.GetTeam(test.team.Id)
+			require.Nil(t, err)
+
+			require.Equal(t, test.wantType, team.Type)
+			require.Equal(t, test.wantOpenInvite, team.AllowOpenInvite)
+		})
+	}
+}
```
Comment (saturninoabril) [inline api4/team_test.go:522]: Can we add tests for the following?
- related to permission (with / without permission to manage a team - forbidden), and
- logged in or not logged in user - unauthorized

File: api4/team.go [hunk 3]
```diff
@@ -231,6 +232,55 @@ func patchTeam(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		openInvite = true
+	case "private":
+		teamType = model.TEAM_INVITE
+		openInvite = false
+	default:
+		c.SetInvalidUrlParam("privacy")
+		return
+	}
+
+	err := c.App.UpdateTeamPrivacy(c.Params.TeamId, teamType, openInvite)
```
Comment (ethervoid) [inline api4/team.go:274]: Maybe `if err := ...; err != nil {}` form?

File: api4/team.go [hunk 4]
```diff
@@ -231,6 +232,55 @@ func patchTeam(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	auditRec := c.MakeAuditRecord("updateTeamPrivacy", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	auditRec.AddMeta("privacy", c.Params.Privacy)
+
+	if !c.App.SessionHasPermissionToTeam(*c.App.Session(), c.Params.TeamId, model.PERMISSION_MANAGE_TEAM) {
+		auditRec.AddMeta("team_id", c.Params.TeamId)
+		c.SetPermissionError(model.PERMISSION_MANAGE_TEAM)
+		return
```
Comment (ethervoid) [inline api4/team.go:253]: One question about the audit system. In this case with the return the audit is marked as failed right?

File: model/client4.go [hunk 5]
```diff
@@ -1794,6 +1794,18 @@ func (c *Client4) PermanentDeleteTeam(teamId string) (bool, *Response) {
 	return CheckStatusOK(r), BuildResponse(r)
 }
 
+// UpdateTeamPrivacy modifies the team type (model.TEAM_OPEN <--> model.TEAM_INVITE) and sets
+// the corresponding AllowOpenInvite appropriately.
+func (c *Client4) UpdateTeamPrivacy(teamId string, privacy string) *Response {
+	query := fmt.Sprintf("?privacy=%s", privacy)
```
Comment (ethervoid) [inline model/client4.go:1800]: In case we move it to the body, change this

File: web/context.go [hunk 6]
```diff
@@ -627,3 +627,14 @@ func (c *Context) RequireBotUserId() *Context {
 	}
 	return c
 }
+
+func (c *Context) RequirePrivacy() *Context {
```
Comment (ethervoid) [inline web/context.go:631]: Given that this parameter is only related to a team, maybe is better to call it `RequireTeamPrivacy`

File: web/context.go [hunk 7]
```diff
@@ -627,3 +627,14 @@ func (c *Context) RequireBotUserId() *Context {
...
+	if c.Err != nil {
+		return c
+	}
+
+	if c.Params.Privacy != "public" && c.Params.Privacy != "private" {
```
Comment (ethervoid) [inline web/context.go:636]: Same here for the constant part

File: api4/team.go (hunk 1 above)
Comment (wiggin77) [reply api4/team.go]: Thanks for that @saturninoabril .  I didn't realize there was a corresponding UpdateChannelPrivacy API.  I've modified this PR to mirror that API.  https://github.com/mattermost/mattermost-server/pull/14287/commits/045457a23d4112cf390b6ae5f244a0ae194f1ec8

File: api4/team_test.go (hunk 2 above)
Comment (wiggin77) [reply api4/team_test.go:530]: Unit tests added here https://github.com/mattermost/mattermost-server/pull/14287/commits/045457a23d4112cf390b6ae5f244a0ae194f1ec8   Thanks!

File: api4/team.go [hunk 8]
```diff
@@ -231,6 +232,55 @@ func patchTeam(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	if team, err := c.App.GetTeam(c.Params.TeamId); err == nil {
+		auditRec.AddMeta("team", team)
+	}
+
+	var teamType string
+	var openInvite bool
+	switch c.Params.Privacy {
+	case "public":
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply api4/team.go]: Thanks @ethervoid .  I modified the PR to be consistent with UpdateChannelPrivacy and now use the existing model.TEAM_OPEN and model.TEAM_INVITE constants throughout.

File: api4/team.go (hunk 3 above)
Comment (wiggin77) [reply api4/team.go]: I made the change.  TBH, I'm not a fan of that pattern when a single function is mixing both styles. Sometimes you need one of the vars elsewhere and must use `something, err := ...`, and then the mixing of style one after the other is harder to read IMO.

File: api4/team.go (hunk 4 above)
Comment (wiggin77) [reply api4/team.go:266]: Correct, the audit record defaults to FAIL until it reaches the auditRec.Success() line.

File: model/client4.go (hunk 5 above)
Comment (wiggin77) [reply model/client4.go]: Done, thanks.  https://github.com/mattermost/mattermost-server/pull/14287/commits/045457a23d4112cf390b6ae5f244a0ae194f1ec8

File: web/context.go (hunk 6 above)
Comment (wiggin77) [reply web/context.go]: I removed this now that the privacy field is moved into the body.

File: web/context.go (hunk 7 above)
Comment (wiggin77) [reply web/context.go]: Done https://github.com/mattermost/mattermost-server/pull/14287/commits/045457a23d4112cf390b6ae5f244a0ae194f1ec8

**Pull request merged**